src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs

```diff
@@ -285,93 +274,102 @@ internal void EntryExpired(CacheEntry entry)
             {
                 ScheduleTask(utcNow);
             }
 
-            void ScheduleTask(DateTimeOffset utcNow)
+            void ScheduleTask(DateTime utcNow)
             {
                 _lastExpirationScan = utcNow;
-                Task.Factory.StartNew(state => ScanForExpiredItems((MemoryCache)state), this,
+                Task.Factory.StartNew(state => ((MemoryCache)state).ScanForExpiredItems(), this,
                     CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default);
             }
         }
 
-        private static void ScanForExpiredItems(MemoryCache cache)
+        private void ScanForExpiredItems()
         {
-            DateTimeOffset now = cache._lastExpirationScan = cache._options.Clock.UtcNow;
+            DateTime utcNow = _lastExpirationScan = UtcNow;
 
-            CoherentState coherentState = cache._coherentState; // Clear() can update the reference in the meantime
+            CoherentState coherentState = _coherentState; // Clear() can update the reference in the meantime
             foreach (KeyValuePair<object, CacheEntry> item in coherentState._entries)
             {
                 CacheEntry entry = item.Value;
 
-                if (entry.CheckExpired(now))
+                if (entry.CheckExpired(utcNow))
                 {
-                    coherentState.RemoveEntry(entry, cache._options);
+                    coherentState.RemoveEntry(entry, _options);
                 }
             }
         }
 
         private bool UpdateCacheSizeExceedsCapacity(CacheEntry entry, CoherentState coherentState)
         {
-            if (!_options.SizeLimit.HasValue)
+            long sizeLimit = _options.SizeLimitValue;
+            if (sizeLimit < 0)
             {
                 return false;
             }
 
-            long newSize = 0L;
             for (int i = 0; i < 100; i++)
             {
                 long sizeRead = coherentState.Size;
-                newSize = sizeRead + entry.Size.Value;
+                long newSize = sizeRead + entry.Size;
 
-                if (newSize < 0 || newSize > _options.SizeLimit)
+                if ((ulong)newSize > (ulong)sizeLimit)
                 {
                     // Overflow occurred, return true without updating the cache size
                     return true;
                 }
 
-                if (sizeRead == Interlocked.CompareExchange(ref coherentState._cacheSize, newSize, sizeRead))
+                long original = Interlocked.CompareExchange(ref coherentState._cacheSize, newSize, sizeRead);
+                if (sizeRead == original)
                 {
                     return false;
                 }
+                sizeRead = original;
             }
 
             return true;
         }
 
         private void TriggerOvercapacityCompaction()
         {
-            _logger.LogDebug("Overcapacity compaction triggered");
+            if (_logger.IsEnabled(LogLevel.Debug))
```

Comment:
Isn't it more effectient to just do the log call if there's no cost in creating the log message? I only ever guard the calls to `LogXXX(...)` methods if producing the log message is costly or allocates.

Reply:
It is significantly cheaper to do the check than call `LogDebug` even if it ends up not logging anything.